File: internal/security/policy.go
// Package security decides whether a call runs: the command safety patterns
// (policy.go) and the tier/mode approval gate (approval.go).
package security

import (
	"errors"
	"fmt"
	"regexp"
	"strings"

	"github.com/overkazaf/re-agent/internal/types"
)

var networkTokens = []string{
	"curl", "wget", "nc", "ncat", "netcat", "nmap", "ssh", "scp", "sftp",
	"rsync", "socat", "openssl s_client", "dig", "whois",
}

// networkPatterns is built once: CommandConcerns runs twice per run_command,
// and compiling fourteen expressions each time is pure waste. Each matches the
// token as a whole word, so `concat_files.sh` does not read as `cat`.
var networkPatterns = func() []namedPattern {
	out := make([]namedPattern, 0, len(networkTokens))
	for _, token := range networkTokens {
		out = append(out, namedPattern{
			label: token,
			re:    regexp.MustCompile(`(?i)(^|[\s;&|])` + regexp.QuoteMeta(token) + `($|[\s;&|])`),
		})
	}
	return out
}()

type namedPattern struct {
	label string
	re    *regexp.Regexp
}

var destructivePatterns = []namedPattern{
	{`/\brm\s+-[^\n;]*r[^\n;]*f\b/i`, regexp.MustCompile(`(?i)\brm\s+-[^\n;]*r[^\n;]*f\b`)},
	{`/\bdd\s+if=/i`, regexp.MustCompile(`(?i)\bdd\s+if=`)},
	{`/\bmkfs\b/i`, regexp.MustCompile(`(?i)\bmkfs\b`)},
	{`/\bdiskutil\s+erase/i`, regexp.MustCompile(`(?i)\bdiskutil\s+erase`)},
	{`/\bshutdown\b/i`, regexp.MustCompile(`(?i)\bshutdown\b`)},
	{`/\breboot\b/i`, regexp.MustCompile(`(?i)\breboot\b`)},
	{`/\blaunchctl\b/i`, regexp.MustCompile(`(?i)\blaunchctl\b`)},
	{`/\bsudo\b/i`, regexp.MustCompile(`(?i)\bsudo\b`)},
	{`/>\s*\/dev\/(?:sd|disk|rdisk)/i`, regexp.MustCompile(`(?i)>\s*/dev/(?:sd|disk|rdisk)`)},
}

var sensitivePatterns = []namedPattern{
	{`/\.ssh\b/i`, regexp.MustCompile(`(?i)\.ssh\b`)},
	{`/\.aws\b/i`, regexp.MustCompile(`(?i)\.aws\b`)},
	{`/\.gnupg\b/i`, regexp.MustCompile(`(?i)\.gnupg\b`)},
	{`/keychain/i`, regexp.MustCompile(`(?i)keychain`)},
	{`/id_rsa/i`, regexp.MustCompile(`(?i)id_rsa`)},
	{`/id_ed25519/i`, regexp.MustCompile(`(?i)id_ed25519`)},
	{`/password/i`, regexp.MustCompile(`(?i)password`)},
	{`/secret/i`, regexp.MustCompile(`(?i)secret`)},
	{`/token/i`, regexp.MustCompile(`(?i)token`)},
}

// CommandConcerns lists everything about a command that deserves a second look,
// in operator-readable form. Empty means unremarkable. Callers decide what to
// do with the list: ask the operator when there is one to ask, refuse otherwise.
func CommandConcerns(command string, policy *types.ExecutionPolicy) ([]string, error) {
	compact := strings.TrimSpace(command)
	if compact == "" {
		return nil, errors.New("command is empty")
	}
	var concerns []string
	for _, pattern := range destructivePatterns {
		if pattern.re.MatchString(compact) {
			concerns = append(concerns, "destructive pattern "+pattern.label)
		}
	}
	if !policy.AllowNetwork {
		for _, pattern := range networkPatterns {
			if pattern.re.MatchString(compact) {
				concerns = append(concerns, fmt.Sprintf("network command '%s' (--allow-network to stop asking)", pattern.label))
			}
		}
	}
	if !policy.AllowSensitive {
		for _, pattern := range sensitivePatterns {
			if pattern.re.MatchString(compact) {
				concerns = append(concerns, fmt.Sprintf("sensitive/secret-like token %s (--allow-sensitive to stop asking)", pattern.label))
			}
		}
	}
	return concerns, nil
}

// ValidateCommand is the hard refusal, for callers with no approval path.
func ValidateCommand(command string, policy *types.ExecutionPolicy) error {
	concerns, err := CommandConcerns(command, policy)
	if err != nil {
		return err
	}
	if len(concerns) > 0 {
		return fmt.Errorf("blocked by policy: %s", strings.Join(concerns, "; "))
	}
	return nil
}

func ValidatePathRead(pathValue string, policy *types.ExecutionPolicy) error {
	if policy.AllowSensitive {
		return nil
	}
	for _, pattern := range sensitivePatterns {
		if pattern.re.MatchString(pathValue) {
			return fmt.Errorf("sensitive path blocked by policy: %s", pathValue)
		}
	}
	return nil
}

func ValidateWriteAllowed(policy *types.ExecutionPolicy) error {
	if !policy.AllowWrites {
		return errors.New("writes are disabled; start with --write to permit write_file")
	}
	return nil
}

File: src/security/policy.ts
import type { ExecutionPolicy } from "../types";

const NETWORK_TOKENS = [
  "curl",
  "wget",
  "nc",
  "ncat",
  "netcat",
  "nmap",
  "ssh",
  "scp",
  "sftp",
  "rsync",
  "socat",
  "openssl s_client",
  "dig",
  "whois",
];

const DESTRUCTIVE_PATTERNS = [
  /\brm\s+-[^\n;]*r[^\n;]*f\b/i,
  /\bdd\s+if=/i,
  /\bmkfs\b/i,
  /\bdiskutil\s+erase/i,
  /\bshutdown\b/i,
  /\breboot\b/i,
  /\blaunchctl\b/i,
  /\bsudo\b/i,
  />\s*\/dev\/(?:sd|disk|rdisk)/i,
];

const SENSITIVE_PATTERNS = [
  /\.ssh\b/i,
  /\.aws\b/i,
  /\.gnupg\b/i,
  /keychain/i,
  /id_rsa/i,
  /id_ed25519/i,
  /password/i,
  /secret/i,
  /token/i,
];

/**
 * Everything about a command that deserves a second look, in operator-readable
 * form. Empty means the command is unremarkable. Callers decide what to do with
 * the list: ask the operator when there is one to ask, refuse otherwise.
 */
export function commandConcerns(command: string, policy: ExecutionPolicy): string[] {
  const compact = command.trim();
  if (!compact) throw new Error("Command is empty.");
  const concerns: string[] = [];
  for (const pattern of DESTRUCTIVE_PATTERNS) {
    if (pattern.test(compact)) concerns.push(`destructive pattern ${pattern}`);
  }
  if (!policy.allowNetwork) {
    const lower = compact.toLowerCase();
    for (const token of NETWORK_TOKENS) {
      if (new RegExp(`(^|[\\s;&|])${escapeRegex(token)}($|[\\s;&|])`, "i").test(lower)) {
        concerns.push(`network command '${token}' (--allow-network to stop asking)`);
      }
    }
  }
  if (!policy.allowSensitive) {
    for (const pattern of SENSITIVE_PATTERNS) {
      if (pattern.test(compact)) concerns.push(`sensitive/secret-like token ${pattern} (--allow-sensitive to stop asking)`);
    }
  }
  return concerns;
}

/** Hard refusal, for callers with no approval path (non-interactive runs). */
export function validateCommand(command: string, policy: ExecutionPolicy): void {
  const concerns = commandConcerns(command, policy);
  if (concerns.length > 0) throw new Error(`Blocked by policy: ${concerns.join("; ")}`);
}

export function validatePathRead(pathValue: string, policy: ExecutionPolicy): void {
  if (policy.allowSensitive) return;
  for (const pattern of SENSITIVE_PATTERNS) {
    if (pattern.test(pathValue)) {
      throw new Error(`Sensitive path blocked by policy: ${pathValue}`);
    }
  }
}

export function validateWriteAllowed(policy: ExecutionPolicy): void {
  if (!policy.allowWrites) {
    throw new Error("Writes are disabled. Start with --write to permit write_file.");
  }
}

function escapeRegex(value: string): string {
  return value.replace(/[.*+?^${}()|[\]\\]/g, "\\$&");
}
